specs/data_structures.md

```diff
@@ -161,23 +174,88 @@ https://developers.google.com/protocol-buffers/docs/proto3
 
 # Hashing
 
-https://en.wikipedia.org/wiki/SHA-3
-
-https://godoc.org/golang.org/x/crypto/sha3
+All protocol-level hashing is done using [Keccak-256](https://keccak.team/keccak.html), and not SHA3-256 ([FIPS 202](https://keccak.team/specifications.html#FIPS_202)). This is to enable compatibility with [Ethereum](https://ethereum.org)'s EVM. Keccak-256 outputs a digest that is 256 bits (i.e. 32 bytes) long.
 
-https://docs.rs/sha3
+Libraries implementing Keccak-256 are available in Go (https://godoc.org/golang.org/x/crypto/sha3) and Rust (https://docs.rs/sha3).
 
 # Signing
 
 
 
-# Merkle Tree
+# Merkle Trees
+
+Merkle trees are used to authenticate various pieces of data across the LazyLedger stack, including transactions, messages, the validator set, etc. This section provides an overview of the different tree types used, and specifies how to construct them.
+
+## Binary Merkle Tree
+
+Binary Merkle trees are constructed in the usual fashion, namely:
+
+For leaf node of message `m`:
+```C++
+v = h(serialize(m))
+```
+
+For internal node with children `l` and `r`:
+```C++
+v = h(l, r) = h(l.v, r.v)
+```
+
+Two exceptions are made, in the case of empty nodes, in order to prevent [CVE-2012-2459](https://nvd.nist.gov/vuln/detail/CVE-2012-2459). The value of an empty leaf node or an empty child node to an internal node is 32-byte zero, i.e. `0x0000000000000000000000000000000000000000000000000000000000000000`.
 
+## Sparse Binary Merkle Tree
 
 
 ## Namespace Merkle Tree
 
+Messages in LazyLedger are associated with a provided _namespace ID_, which identifies the application (or applications) that will read these messages when parsing blocks. The Namespace Merkle Tree (NMT) is a variation of the [Merkle Interval Tree](https://eprint.iacr.org/2018/642), which is itself an extension of the [Merkle Sum Tree](https://bitcointalk.org/index.php?topic=845978.0).
+
+Construction of a NMT is similar to that of a plain binary Merkle tree, but with a different hashing method that commits to intervals of namespace IDs.
+
+For leaf node of message `m`:
+```C++
+n_min = m.namespace_id
+n_max = m.namespace_id
+v = h(serialize(m))
+```
+
+The `namespace_id` field is the namespace ID of the message, which is a [`NAMESPACE_ID_BYTES`](consensus.md#system-parameters)-byte unsigned integer.
+
+Before being hashed, the [message](#message)s are [serialized](#serialization).
+
+For internal node with children `l` and `r`:
+```C++
+n_min = min(l.n_min, r.n_min)
+n_max = max(l.n_max, r.n_max)
+v = h(l, r) = h(l.min, l.max, l.v, r.min, r.max, r.v)
```

Comment:
Maybe my above comment doesn't matter. As namespacing effectively differentiates between leaf and inner nodes, too.

Reply:
Yes, I'm pretty sure that you can't execute the same second-preimage attack here.